RFDs/0001-job-worker-service.md

```diff
@@ -0,0 +1,406 @@
+- Resource control using cgroups (CPU, Memory, Disk I/O)
+- Secure communication using mTLS
+- Simple role-based authorization scheme
+- gRPC API for client-server communication
+
+#### Key Considerations 
+- This design is not a distributed system and is only optimized to run on a single linux OS
+- If needed, the components of the library can be split into a distributed microservices framework inthe future
+
+## Detailed Design
+
+### 1. Library
+
+#### a. In-memory Job Logger
+- Implement a thread-safe buffer to store output from runnning the command
+- Use a read-write mutex to ensure thread-safety for concurrent read/write operations
+- Provide methods for adding logs, and getting a stream Reader to the log
+
+```go
+// Logger represents an in-memory buffer for storing job logs.
+
+type Logger struct {
+	mu   sync.RWMutex
+	logs bytes.Buffer
+}
+
+// NewLogger creates a new Logger instance
+func NewLogger() *Logger
+
+// AddLog appends new output from processes to the logs buffer
+func (l *Logger) AddLog(output []byte)
+
+// Gets an instance of io.Reader that is able to read the data from the internal logs buffer
+func (l *Logger) GetLogReader() *io.Reader
+
+// Size returns the current size of the logs buffer
+func (l *Logger) Size() int
+
+```
+
+#### b. Job Executor
+- Use `os/exec` package to create and manage child processes
+- Implement non-blocking I/O for stdout/stderr capture using `io.Pipe()`
+- Use a separate goroutine for each of stdout and stderr to prevent blocking
+- Implement a context-based cancellation mechanism for graceful job termination
+- Monitor resource usage using cgroups
+- Provide callbacks for resource usage updates
+- Use a Go channel to periodically read current resource usage (memory, I/O, and CPU usage) from the associated cgroup
+
+```go
+// Executor is responsible for running a job and managing its lifecycle.
+
+// JobStatus represents the current state of the job
+type JobStatus string
+
+const (
+	JobStatusPending   JobStatus = "PENDING"
+	JobStatusRunning   JobStatus = "RUNNING"
+	JobStatusCompleted JobStatus = "COMPLETED"
+	JobStatusFailed    JobStatus = "FAILED"
+	JobStatusStopped   JobStatus = "STOPPED"
+)
+
+type Executor struct {
+	Command   string
+	Args      []string
+	cmd     *exec.Cmd
+	logger  *logger.Logger
+	resourceManager *resourcemanager.ResourceManager
+	// Context to propagate state and cancellations
+	ctx     context.Context
+	cancel  context.CancelFunc
+	started time.Time
+	stopped time.Time
+	mu      sync.Mutex
+	status  JobStatus
+}
+
+// NewExecutor creates a new Executor for the given command and arguments.
+// cgroupPath will be used to set the path of the cgroup relative to the root cgroup -> /sys/fs/cgroup/jobworker
+func NewExecutor(command string, args []string, cgroupPath string) *Executor
+
+// Start begins the execution of the job.
+func (e *Executor) Start() error
+
+// Stop terminates the running job.
+func (e *Executor) Stop() error
```

Comment:
How are processes stopped? If a long-running process spawns child processes, how are all of the processes gracefully terminated? What if they don't quit after a period of time?